### ISSUE TITLE
Possibly unexpected warning when EnsuresNotNull attribute is involved

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#29865]
```
        [Fact]
        public void EnsuresNotNull_BackEffect()
        {
            CSharpCompilation c = CreateCompilation(new[] { @"
using System.Runtime.CompilerServices;
public class C
{
    void M(string? s1, string? s2)
    {
        ThrowIfNull(s2 = s1, s1);
        s2.ToString(); // warn
    }
    public static void ThrowIfNull(string? x1, [EnsuresNotNull] string? x2) => throw null;
}
", EnsuresNotNullAttributeDefinition, NonNullTypesTrue, NonNullTypesAttributesDefinition });

            // PROTOTYPE(NullableReferenceTypes): Should we be able to trace that s2 was assigned a non-null value?
            c.VerifyDiagnostics(
                // (8,9): warning CS8602: Possible dereference of a null reference.
                //         s2.ToString(); // warn
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s2").WithLocation(8, 9)
                );
        }


        [Fact]
        public void EnsuresNotNull_ForwardEffect()
        {
            CSharpCompilation c = CreateCompilation(new[] { @"
using System.Runtime.CompilerServices;
public class C
{
    void M(string? s1, string? s2)
    {
        ThrowIfNull(s1, s2 = s1);
        s2.ToString(); // ok
    }
    public static void ThrowIfNull([EnsuresNotNull] string? x1, string? x2) => throw null;
}
", EnsuresNotNullAttributeDefinition, NonNullTypesTrue, NonNullTypesAttributesDefinition });

            c.VerifyDiagnostics();
        }
```
